source/common/router/router.cc

```diff
@@ -1329,6 +1329,19 @@ bool Filter::maybeRetryReset(Http::StreamResetReason reset_reason,
                           ? RetryState::Http3Used::Yes
                           : RetryState::Http3Used::No;
   }
+
+  auto upstream_request_started = false;
+  // If any request in this router has sent data to the upstream, we consider the request started.
+  for (auto& request : upstream_requests_) {
+    if (request->streamInfo()
+            .upstreamInfo()
+            ->upstreamTiming()
+            .first_upstream_tx_byte_sent_.has_value()) {
+      upstream_request_started = true;
+      break;
+    }
+  }
```

Comment:
is necessary to iterate the list every time? And note when you try to do the third times retry, the first upstream request may had been removed from the list.

So, I think we can just add a bool flag to the `Router::Filter` with initial value `false`.

Then to update the value here by
```
upstream_request_started_ |= upstream_request.streamInfo().upstreamInfo()->upstreamTiming().first_upstream_tx_byte_sent_.has_value();
```

Reply:
The first version of this PR had a bool flag but @alyssawilk wanted to avoid that if possible. Can you elaborate on why the first request may be removed on a third retry?

Reply:
See, assume we set the max retry to 5 with multiple different retry condition. Then when you failed three time and try to emit your fourth upstream request, it's no double the initial upstream request was removed. See

```
  if (retry_status == RetryStatus::Yes) {
    runRetryOptionsPredicates(upstream_request);
    pending_retries_++;
    if (upstream_request.upstreamHost()) {
      upstream_request.upstreamHost()->stats().rq_error_.inc();
    }
    auto request_ptr = upstream_request.removeFromList(upstream_requests_);
    callbacks_->dispatcher().deferredDelete(std::move(request_ptr));
    return true;
```

I may have different view on this point with @alyssawilk . let me take a quick check to the history context.

Reply:
Hmmm, I will prefer to keep the bool flag and no the new method/interface. And you can update the flag as my previous suggestion.

Reply:
Added!